### ISSUE TITLE
Getting pixels

### ISSUE THREAD

**Issue by giuseb** [minimagick/minimagick#365]
I switched from RMagick to minimagick exactly for the reasons you stated in the README (kudos to you!)

The only thing I truly miss from RMagick is the ability to `img.get_pixels()`.
Since I do need that functionality, I ended up putting together something like the following (simplified for grayscale pixels):

``` ruby
path = '/path/to/my/image.jpg'
pixels = IO.read("|convert #{path} gray:-").unpack 'C*'
# reshape array according to image size
# (although in reality I would use NArray or NMatrix)
width = IO.read("|identify -format '%w' #{path}").to_i
matrix = pixels.each_slice(width).to_a
```

Do you think it might be worth implementing this type of behavior, or is this outside of minimagick's scope?

Thanks, Giuseppe

**Comment by janko-m** [comment]
I'm up for adding `MiniMagick::Image#get_pixels`. RMagick seems to be using `AcquireImagePixels` from ImageMagick's C bindings, I will search if something like this is exposed in the command-line. Otherwise I will try with a modified version of your code (thank you for sharing it).

**Comment by janko-m** [comment]
@giuseb Hmm, I don't really know how to write this method, I don't get the correct result 🤔

So, I tried this:

```rb
    def get_pixels(*args)
      convert = MiniMagick::Tool::Convert.new
      if args.any?
        raise ArgumentError, "must provide 4 arguments: (x, y, columns, rows)" if args.size != 4
        x, y, columns, rows = args
        convert << "#{path}[#{x}x#{y}+#{columns}+#{rows}]"
      else
        columns = width
        convert << path
      end
      convert.depth(8)
      convert << "RGB:-"
      content = convert.call

      pixels = content.unpack("C*")
      pixels.each_slice(columns).to_a
    end
```

I would now expect that the total number of pixels would equal `width * height`, but it doesn't, the number of pixels is much bigger. Am I incorrectly evaluating the correctness of the result? Also, what kind of numbers should the pixels be?

**Comment by Coolnesss** [comment]
Hey, thanks a lot to both of you for your codes, they were very useful for me. 

@janko-m I was able to make you code work (I think), a bare-bones version looks like the following. It only supports getting the entire image pixels though.

```ruby
module MiniMagick
  class Image
    def get_pixels
      convert = MiniMagick::Tool::Convert.new
      convert << path
      convert.depth(8)
      convert << "RGB:-"
      content = convert.call
      pixels = content.unpack("C*")
      pixels.each_slice(3).map{|x| x.sum / 3.0}
    end
  end
end
```

I think the trick was that each pixel has three values, red green and blue, and thus my code takes the average of those to get the value of one pixel. You can obviously change this in the code. The dimensions seem to match now, as a 400x480 image yields a 192000 size result, which is the value of the product 400*480.

Let me know if i can be of help otherwise, i'd be glad if the code (in some form) got merged.

**Comment by giuseb** [comment]
@janko-m: in my code, the image is converted to grayscale, where each pixel is represented by one unsigned 8-bit integer (which is what [unpacking](http://apidock.com/ruby/String/unpack) ensures). Thus, the size of the resulting matrix is column-by-rows bytes.

As @Coolnesss correctly points out, the full RGB image is 3 times as large, as each pixel is described by 3 bytes, one for each color channel.

@Coolnesss: by computing the average value of each pixel, you are again converting to grayscale. In addition, your `get_pixels()` returns a single long row of pixels instead of a rectangular matrix.

In summary, a `get_pixels()` method that returns RGB values should end with something like:

    pixels.each_slice(3).each_slice(width).to_a

As a result:

    p = get_pixels   # use arguments to crop 
    row = 0          # get the first row
    col = 1          # get the second column
    cha = 1          # get the green channel
    p[row][col][cha] # the looked-for value

**Comment by Coolnesss** [comment]
@giuseb yes sir I understand that. 

I'll draft up a pull request when i have some time. I'd like the method to work similarly to the rmagick method, where one can specify a rectangle inside the image as well.

**Comment by giuseb** [comment]
Hello, I have pushed a [new branch on my own fork](https://github.com/giuseb/minimagick/tree/get-pixels), hoping it may be useful as a starting point for a full-featured method.

A couple of notes:

@janko-m, for reasons I do not understand, the compact syntax you were using to extract a portion of the image does not work as expected (all pixels from the original image are retrieved, instead of just the subset). Using a "regular" `crop` command seems to do the trick.

Also, note that ImageMagick's image geometry dictates that columns and rows are specified **before** x and y offsets (as in "CxR+X+Y").

BTW, wouldn't something like `get_pixels("400x300+20+10")` be more in line with  minimagick's style?

**Comment by Coolnesss** [comment]
Hey @giuseb actually I just made something very simliar to yours, but I didn't see your code before. Have a look. 

I also thought maybe something that directly uses ImageMagick syntax would be better, but on the other hand this library is a lightweight version of RMagick, so RMagick syntax seems like a good choice too.

**Comment by giuseb** [comment]
Hey @Coolnesss, the more I think about it, the more it seems to me that the IM-like image-geometry string is the better option to specify a subregion in `get_pixels`. It requires significantly less code on our end, since it leverages IM's parser and error checking.

I have pushed a modified version of the method, and I am going to open a separate PR (hope you don't mind, I don't mean to be bossy :-)

Let me know what you think!

**Comment by Coolnesss** [comment]
@giuseb Looks good, no problem! I'll close mine, your code looks much better.

**Comment by janko-m** [comment]
https://github.com/minimagick/minimagick/pull/393 was merged, so closing this.
